### PR TITLE
refactor: sending methods, UI

### LINKED ISSUE #1067: Can't send STX from non-first account
Seems to be a bug where if you try to send STX from an account index > 0 it breaks.

Steps:

- make new account
- switch to it
- have stx
- try to do a token transfer
- errors with something around BN

Screen Shot 2021-03-12 at 5 13 42 PM

### PR DESCRIPTION
**Pull request by aulneau**
> Try out this version of the Stacks Wallet - download [extension builds](https://github.com/blockstack/ux/actions/runs/691420206).

### What this does

This PR fixes https://github.com/blockstack/stacks-wallet-web/issues/1067. It seems that there was some issues with the way that we were keeping `StacksTransactions` in recoil store. Recoil serializes everything that is in an atom/selector, and I think that serialization was breaking the transaction class.

#### Changes & Improvements
- validation has been improved on the send screen
- send screen design has been improved slightly moving towards the figma designs
- tickers are now displayed in the same way as the explorer
- error handling now displays a toast if the transaction fails for some reason
- assets now use the same kind of gradient as on the explorer
- amount placeholder updates based on asset selected

stx-transfer

stella-transfer

### CODE REVIEW

File: src/common/hooks/use-asset-transfer.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+interface PostConditionsOptions {
+  contractAddress: string;
+  contractName: string;
+  assetName: string;
+  stxAddress: string;
+  amount: number;
+}
+
+function makePostConditions(options?: PostConditionsOptions) {
```
Comment (kyranjamie) [inline src/common/hooks/use-asset-transfer.ts]: Any reason not to throw an error here? When would you want to make a post condition without options?

File: src/common/hooks/use-make-stx-transfer.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+export function useMakeStxTransfer() {
+  return useRecoilCallback(({ snapshot }) => async (params: TokenTransferParams) => {
+    const { amount, recipient } = params;
+    const stacksNetwork = await snapshot.getPromise(stacksNetworkStore);
+    const account = await snapshot.getPromise(currentAccountStore);
+    const nonce = await snapshot.getPromise(correctNonceStore);
+
+    if (account) {
+      return await makeSTXTokenTransfer({
```
Comment (kyranjamie) [inline src/common/hooks/use-make-stx-transfer.ts]: unneeded await?

File: src/common/hooks/use-make-stx-transfer.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  const { isLoading, setIsLoading, setIsIdle } = useLoading(loadingKey);
+  const asset = useRecoilValue(selectedAssetStore);
+  const handleMakeStxTransaction = useMakeStxTransfer();
+  const handleMakeFtTransaction = useMakeAssetTransfer();
+  const isActive = isShowing && amount && recipient;
+  const notLoaded = !transaction && !isLoading;
+
+  useEffect(() => {
+    const method = asset?.type === 'stx' ? handleMakeStxTransaction : handleMakeFtTransaction;
+    if (isActive) {
```
Comment (kyranjamie) [inline src/common/hooks/use-make-stx-transfer.ts:61]:
```suggestion
    if (isActive && notLoaded) {
```
?

File: src/components/asset-row.tsx [hunk 4]
```diff
@@ -1,42 +1,49 @@
...
+      alignItems="center"
+      flexWrap="wrap"
+      flexDirection="row"
+      cursor="pointer"
+      width="100%"
+      {...rest}
+      ref={ref}
+    >
+      <AssetAvatar
+        useStx={name === 'STX' || name === 'Stacks Token'}
```
Comment (kyranjamie) [inline src/components/asset-row.tsx:29]: useStx seems like a strange prop name?

File: src/components/drawer/confirm-send-drawer.tsx [hunk 5]
```diff
@@ -1,121 +1,132 @@
...
+    amount: number;
+    nonce?: number;
+    fee?: number;
+    recipient: string;
+  } & StackProps
+> = ({ amount, nonce, fee, recipient, ...rest }) => {
   const asset = useRecoilValue(selectedAssetStore);
-  const stacksNetwork = useRecoilValue(stacksNetworkStore);
-  const { doChangeScreen } = useAnalytics();
+  const ticker = asset && asset.type !== 'stx' ? getTicker(asset.name).toUpperCase() : 'STX';
```
Comment (kyranjamie) [inline src/components/drawer/confirm-send-drawer.tsx]: `isStx` might be a nice helper

**Review by kyranjamie — APPROVED** [review]

File: src/common/hooks/use-asset-transfer.ts (hunk 1 above)
Comment (hstove) [reply src/common/hooks/use-asset-transfer.ts]: Agreed - also, it seems odd to call this `makePostConditions`, but it only takes options for one PC, and returns an array of one PC. I know we are only making on PC for internal transfers. Still, I think this should just be called `makePostCondition`, and take mandatory options, and returns one PC.

File: src/common/hooks/use-asset-transfer.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+          contractName,
+          assetName,
+          stxAddress,
+          amount,
+        }
+      : undefined;
+
+    const postConditions = makePostConditions(postConditionOptions);
+
+    const functionArgs = [standardPrincipalCVFromAddress(createAddress(recipient)), uintCV(amount)];
```
Comment (hstove) [inline src/common/hooks/use-asset-transfer.ts]: The signature has changed in the new SIP:

```
(transfer (uint principal principal) (response bool uint))
```

So the args are `amount, sender, recipient`. We should just use that new signature now that we're making this change.

File: src/common/hooks/use-make-stx-transfer.ts (hunk 2 above)
Comment (hstove) [reply src/common/hooks/use-make-stx-transfer.ts]: It is an async method - it fetches the nonce and fee rate, or in this case just the fee rate.

File: src/common/hooks/use-make-stx-transfer.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import { makeSTXTokenTransfer, StacksTransaction } from '@stacks/transactions';
+import BN from 'bn.js';
+import { stxToMicroStx } from '@stacks/ui-utils';
+import { useLoading } from '@common/hooks/use-loading';
+import { useEffect } from 'react';
+import { useMakeAssetTransfer } from '@common/hooks/use-asset-transfer';
+import { selectedAssetStore } from '@store/recoil/asset-search';
+
+interface TokenTransferParams {
+  amount: number;
```
Comment (hstove) [inline src/common/hooks/use-make-stx-transfer.ts:14]: I'm a little wary in using `number` for all the `amounts` around this. Strings are just safer. We only need to convert to a `number` when validating the user's balance - and even then, a BN-type object is better

File: src/common/utils.ts [hunk 8]
```diff
@@ -162,3 +162,19 @@ export function truncateString(str: string, maxLength: number) {
...
+  if (name.includes('-')) {
+    const parts = name.split('-');
+    if (parts.length >= 3) {
+      return `${parts[0][0]}${parts[1][0]}${parts[2][0]}`;
+    } else {
+      return `${parts[0][0]}${parts[1][0]}${parts[1][1]}`;
+    }
+  } else {
+    if (name.length >= 3) {
+      return `${name[0]}${name[1]}${name[2]}`;
```
Comment (hstove) [inline src/common/utils.ts]: `name.slice(0, 3)` is cleaner

File: src/common/utils.ts [hunk 9]
```diff
@@ -162,3 +162,19 @@ export function truncateString(str: string, maxLength: number) {
   }
   return str.slice(0, maxLength) + '...';
 }
+
+export function getTicker(name: string) {
```
Comment (hstove) [inline src/common/utils.ts]: Maybe add a comment with some example examples. This is also a great place to add some simple unit tests

File: src/components/asset-row.tsx (hunk 4 above)
Comment (hstove) [reply src/components/asset-row.tsx:29]: Yeah - this is likely a result of code I wrote, but it also seems smelly to have two possible `name`s for a STX asset.

File: src/components/asset-search/selected-asset.tsx [hunk 10]
```diff
@@ -34,14 +38,15 @@ export const SelectedAsset: React.FC<{ hideArrow?: boolean }> = ({ hideArrow })
   if (!selectedAsset) {
     return null;
   }
-  const { name } = selectedAsset;
+  const { name, contractAddress } = selectedAsset;
+
+  const isStx = name === 'Stacks Token';
+  const ticker = isStx ? 'STX' : getTicker(name).toUpperCase();
```
Comment (hstove) [inline src/components/asset-search/selected-asset.tsx]: Another place where these names end up with smelly code. Probably the better solution is to have Asset have two interfaces - StxAsset and FungibleAsset, then type Asset = StxAsset | FungibleAsset

Comment (hstove) [reply src/components/asset-search/selected-asset.tsx]: Also, is there every a reason that `getTicker` should not always return full caps? Then we don't need to `toUpperCase` it always?

File: src/components/drawer/confirm-send-drawer.tsx (hunk 5 above)
Comment (hstove) [reply src/components/drawer/confirm-send-drawer.tsx]: Another example for `getTicker` to return full caps

**Comment by hstove** [comment]
Alignment of asset icons in asset list seems off - they're bottom aligned but should be vertically aligned, yeah?

image

File: src/pages/popup/index.tsx [hunk 11]
```diff
@@ -53,39 +54,50 @@ const CopyAction: React.FC<BoxProps> = memo(({ onClick }) => {
...
   const ref = useRef<HTMLButtonElement | null>(null);
+  const { doChangeScreen } = useAnalytics();
+  const assets = useAssets();
+
+  const isSend = kind === 'send';
+  const sendDisabled = isSend && assets.length === 0;
+
+  const handleClick = useCallback(() => {
+    doChangeScreen(path);
+  }, [path, doChangeScreen]);
```
Comment (hstove) [inline src/pages/popup/index.tsx:67]: Maybe just delete the `path` option and have this handler simple pick the path based on `kind`? Mainly a nit

File: src/store/recoil/loading.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+import { atomFamily } from 'recoil';
+
+export const loadingState = atomFamily<'idle' | 'loading', string>({
```
Comment (hstove) [inline src/store/recoil/loading.ts:3]: I likey, but personally might prefer a simple boolean? Nit

File: package.json [hunk 13]
```diff
@@ -50,7 +50,7 @@
     "@stacks/connect-ui": "^3.0.2",
     "@stacks/network": "^1.0.2",
     "@stacks/rpc-client": "^1.0.3",
-    "@stacks/transactions": "1.3.0",
+    "@stacks/transactions": "^1.3.0",
```
Comment (hstove) [inline package.json]: We need to keep these pinned (which is why there are conflicts), but also after you rebase the new CI will flag this

**Review by hstove — APPROVED** [review]
Generally :+1:, would appreciate some small cleanup changes

File: src/components/asset-row.tsx (hunk 4 above)
Comment (aulneau) [reply src/components/asset-row.tsx:29]: I agree, this was there already and I think we should punt on fixing it. This issue is not the goal of the PR. When the UI is updated, it's likely this component will be completely refactored.

File: src/common/hooks/use-make-stx-transfer.ts (hunk 2 above)
Comment (kyranjamie) [reply src/common/hooks/use-make-stx-transfer.ts]: Yes, but it's returning to an async function

**Pull request merged**